persistent/Database/Persist/TH/Internal.hs

```diff
@@ -282,7 +282,11 @@ embedEntityDefsMap existingEnts rawEnts =
 --
 -- @since 2.5.3
 parseReferences :: PersistSettings -> [(Maybe SourceLoc, Text)] -> Q Exp
-parseReferences ps s = lift $ parse ps s
+parseReferences ps s = do
+  let cpr = parse ps s
+  case renderErrors cpr of
+    Nothing -> lift $ cumulativeData cpr
+    Just errorMessage -> fail errorMessage
```

Comment:
So here's where `CumulativeParseResult` is finally consumed. We use `renderErrors` to determine if there are any errors, and if not, we `lift` the cumulative data. If there are errors, we fail with error message.

I think we probably want to represent `CumulativeParseResult` instead as a `Either (NonEmpty Error) a`. Then the resulting logic is all much simpler.